java-checks-test-sources/src/main/java/checks/BoxedBooleanExpressionsCheck.java

```diff
@@ -341,4 +341,72 @@ Boolean False() {
 
   Object foo() { return new Object(); }
   Object bar() { return new Object(); }
+
+  void forStatementCheckedBeforeUseIsCompliant() {
+    Boolean checkedBeforeUse = getCustomNonNull();
+    if (checkedBeforeUse != null) {
+      for (; checkedBeforeUse; )  {} // Compliant because the variable is checked before use
+    }
+
+    Boolean checkedBeforeUse2 = getCustomNonNull();
+    if (checkedBeforeUse2 == null) {
+      for (; checkedBeforeUse2; )  {} // Compliant because the variable is checked before use
+    }
+  }
+
+  void whileStatementCheckedBeforeUseIsCompliant() {
+    Boolean checkedBeforeUse = getCustomNonNull();
```

Comment:
I think you don't need to use <code>getCustomNonNull()</code>, but sth like <code>customNullable</code> because the main reason to use @NonNull annotations is to get rid of null checks. And I think the cases with annotation are already covered.